src/Main.elm

```diff
@@ -1,10 +1,418 @@
+import Data.AudienceFolder exposing (AudienceFolder)
+import Data.Store as Store exposing (AudienceFolderID, AudienceLevel, Store)
+import Element exposing (Element, column, el, none, paragraph, row, text)
+import Element.Background
+import Element.Border
+import Element.Events
+import Element.Font
+import Element.Input exposing (button)
+import Html
+import Html.Attributes
+import Http
+import Task
 
 
-main : Html msg
+
+---------------
+-- M O D E L --
+---------------
+
+
+type alias InitialisingState =
+    { folders : Maybe (List AudienceFolder)
+    , audiences : Maybe (List Audience)
+    }
+
+
+type Location
+    = Root
+    | Folder AudienceFolderID
+
+
+type alias SucceedState =
+    { location : Location
+    , filterBy : AudienceType
+    }
+
+
+type Model
+    = Initialising InitialisingState
+    | Failed Http.Error
+    | Succeed Store SucceedState
+
+
+makeSelector : SucceedState -> Store.Selector
+makeSelector state =
+    case state.filterBy of
+        Audience.Authored ->
+            case state.location of
+                Root ->
+                    Store.OnlyAuthored
+
+                Folder folderID ->
+                    Store.OnlyAuthoredIn folderID
+
+        Audience.Shared ->
+            Store.OnlyShared
+
+        Audience.Curated ->
+            case state.location of
+                Root ->
+                    Store.OnlyCurated
+
+                Folder folderID ->
+                    Store.OnlyCuratedIn folderID
+
+
+turnInitialising : InitialisingState -> Model
+turnInitialising state =
+    case Maybe.map2 Store.init state.folders state.audiences of
+        Nothing ->
+            Initialising state
+
+        Just store ->
+            Succeed store (SucceedState Root Audience.Authored)
+
+
+init : flags -> ( Model, Cmd Msg )
+init _ =
+    ( Initialising
+        { folders = Nothing
+        , audiences = Nothing
+        }
+    , Cmd.batch
+        [ Task.attempt LoadAudienceFoldersDone Api.getListOfAudienceFolders
+        , Task.attempt LoadAudiencesDone Api.getListOfAudiences
+        ]
+    )
+
+
+
+-----------------
+-- U P D A T E --
+-----------------
+
+
+type Msg
+    = LoadAudienceFoldersDone (Result Http.Error (List AudienceFolder))
+    | LoadAudiencesDone (Result Http.Error (List Audience))
+    | GoDown AudienceFolderID
+    | GoUp
+    | SetFilterBy AudienceType
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case ( msg, model ) of
+        ( LoadAudienceFoldersDone (Err err), _ ) ->
+            ( Failed err, Cmd.none )
+
+        ( LoadAudienceFoldersDone (Ok listOfFolders), Initialising state ) ->
+            ( turnInitialising { state | folders = Just listOfFolders }
+            , Cmd.none
+            )
+
+        ( LoadAudienceFoldersDone _, _ ) ->
+            ( model, Cmd.none )
+
+        ( LoadAudiencesDone (Err err), _ ) ->
+            ( Failed err, Cmd.none )
+
+        ( LoadAudiencesDone (Ok listOfAudiences), Initialising state ) ->
```

Comment:
To be honest this whole approach seems to me too complicated and hard to understand what is covered and what is not. Yes, you have a compiler to say you, what case has to be covered, but in this line is an example of a false sense of security. 
Image when you will have the ability to reload audiences/folders even if they are already loaded. 
Someone who did not write this code will use `Api.getListOfAudiences` together with existing Msg `LoadAudiencesDone` and WTF moment is here. You can say, it's no big issue, but when you have huge codebase full of those hidden non-obvious logic, it's always pain.

Reply:
Here is elm app https://app.texenergo.com with 44k lines of elm code in 109 files which include routing and handling 36 url routes. Still no problem with the approach.

Reply:
At the moment of implementing the logic there was not ability to load the data when the app already have it. Make impossible state impossible.